### DISCUSSION TITLE
New APT and YUM repository and publishing system

### DISCUSSION THREAD

**Discussion by FooBarWidget** [fullstaq-labs/fullstaq-ruby-server-edition#78] [Show and tell]
Today we launched [a new APT and YUM repository](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/blob/main/dev-handbook/apt-yum-repo.md) for Fullstaq Ruby, as well as a new CI/CD process for testing and publishing packages.

This is a big deal because Bintray — on which we hosted our packages — sunsetted on March 31.

We found that there are no good hosted alternatives to Bintray, which meant that we needed to develop an alternatives ourselves. But building an APT/YUM repository and a corresponding publishing process is actually very difficult and a lot of work, riddled with many caveats. Just putting up a repository is easy; making the repository and publishing process *reliable, highly available, consistent, secure and playing well with CI/CD* is the hard part.

After many weeks of hard work, this effort is now finally done. In the following technical document, I describe how our APT/YUM repository is set up and how the publishing process works. Special attention was given to the following problems:
- Atomic publishing
- Strong consistency
- Concurrency safe publishing
- Segregated, temporary repositories for testing
- Production data parity during testing
- Preservation of old package versions
- Optional overwrite protection
- Backups

This could be a good inspiration for others looking to setup a reliable APT/YUM repository and publishing process.

Read all about it here: [APT and YUM repository setup](https://github.com/fullstaq-labs/fullstaq-ruby-server-edition/blob/main/dev-handbook/apt-yum-repo.md)

**Comment by mperham** [comment]
I’ve done about 20% of this for my own products and it’s a huge PITA. Thank you Hongli and team for this amazing work. 😎
